### PR TITLE
add support for in-memory configuration to address #1649

### PR DESCRIPTION
**Pull request by drochetti**
*Issue #, if available:* #1649

*Description of changes:*

- added `AWSInfo.configure` so configuration can come from a
`NSDictionary` as an alternative to the default `awsconfiguration.json`
- added tests to make sure when `AWS.configure` is used, that will take
precedence over the JSON file
- added tests to verify if the configuration was set correctly

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Nikooos** [comment]
Would love to see this PR merged soon!

**Comment by JesusMartinAlonso** [comment]
I hope this PR is merged soon. The configuration through a .json file in the bundle is very insecure: If you uncompress the .ipa file you can retrieve the file with all the keys :(

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,15 @@
...
 
+### New Features
+
+- **AWSCore**
+  - Added the option of passing the configuration as an in-memory object (i.e. `[String: Any]`/`NSDictionary`) instead of the default `awsconfiguration.json` through the new API `AWSInfo.configureDefaultAWSInfo(config: [String: Any])`.
+- **AWSMobileClient**
+  - Based on the ability to pass a custom configuration through an in-memory object described above, exposed a new initializer that accepts a custom configuration: `AWSMobileClient.init(configuration: [String: Any])`.
+  - Changed `AWSMobileClient.sharedInstance()` in favor of `AWSMobileClient.default()` since it better communicates the API intent. The `sharedInstance` is still available for backwards compatibility.
+  - Refer to [Issue #1649](https://github.com/aws-amplify/aws-sdk-ios/issues/1649) for the feature request details.
+
```
Comment (palpatim) [inline CHANGELOG.md]: Add a note that you're removing deprecated items (see @royjit's recent PR for suggested wording)

File: AWSAuthSDK/Sources/AWSMobileClient/Internal/_AWSMobileClient.m [hunk 2]
```diff
@@ -64,16 +64,31 @@ + (instancetype)sharedInstance {
...
     
     return _sharedMobileClient;
 }
 
+#pragma initializers
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        self.isInitialized = NO;
```
Comment (royjit) [inline AWSAuthSDK/Sources/AWSMobileClient/Internal/_AWSMobileClient.m]: Better not to use `self.isInitialized` and `self.signInProviderConfig` inside the init method. On init the object is not fully initialized and calling the accessors can cause side effects.

File: CHANGELOG.md (hunk 1 above)
Comment (drochetti) [reply CHANGELOG.md]: I spoke to @royjit and he will merge his PR (https://github.com/aws-amplify/aws-sdk-ios/pull/1738) so we keep the deprecation isolated from the in-memory config changes.

File: AWSAuthSDK/Sources/AWSMobileClient/Internal/_AWSMobileClient.m (hunk 2 above)
Comment (drochetti) [reply AWSAuthSDK/Sources/AWSMobileClient/Internal/_AWSMobileClient.m]: Thanks, that's a good call! I changed it and will publish a new revision soon

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift [hunk 3]
```diff
@@ -160,11 +159,21 @@ final public class AWSMobileClient: _AWSMobileClient {
     
     /// The singleton instance of `AWSMobileClient`.
     ///
+    /// **Note:** prefer to use `AWSMobileClient.default()` since it communicates better the API intent.
+    /// You can manage your own instance of `AWSMobileClient` through the `AWSMobileClient.init(configuration)`.
+    ///
     /// - Returns: The singleton `AWSMobileClient` instance.
```
[the remark this answers is not in the document]
Comment (drochetti) [reply AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift]: I went back and forth on this but I agree, marking as deprecated offer the consumers a self-documented and easy fix (also Xcode works pretty well when the `rename` property is provided... neat!). My next revision will contain the proposed change

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift [hunk 4]
```diff
@@ -169,7 +169,14 @@ final public class AWSMobileClient: _AWSMobileClient {
...
     ///
-    /// You can manage your own instance of `AWSMobileClient` through the `AWSMobileClient.init(configuration)`.
+    /// You can manage your own instance of `AWSMobileClient` by constructing it with `AWSMobileClient(configuration)`,
+    /// however please note that multiple instances of AWSMobileClient is not supported.
+    ///
+    /// **Implementation Notes:**
+    ///
+    /// The `AWSMobileClient` relies on `AWSInfo` for configuration and once that class is initialized, the configuration
+    /// cannot be reset and/or re-initialized. Therefore, even though you can instantiate `AWSMobileClient` multiple times,
+    /// all instances will have teh same configuration reference.
```
Comment (palpatim) [inline AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift]: s/teh/the/ :)

Comment (drochetti) [reply AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift]: oops! thanks for catching it

**Comment by drochetti** [comment]
> I hope this PR is merged soon. The configuration through a .json file in the bundle is very insecure: If you uncompress the .ipa file you can retrieve the file with all the keys :(

Hi @JesusMartinAlonso, since this is a big change we were going through several reviews and testing cycles to make sure it is correct and backwards compatible. The PR is now merged into the `develop` branch and will go out in the `2.11.0` release that should happen soon. I appreciate your patience. =)

Regarding your concern with the `awsconfiguration.json` file, we have [a section in our documentation](https://aws-amplify.github.io/docs/cli-toolchain/usage#awsconfigurationjson-file) that should address it.

**Comment by palpatim** [comment]
We released this in https://github.com/aws-amplify/aws-sdk-ios/releases/2.11.0

**Comment by JesusMartinAlonso** [comment]
@pokryfka  I have the same problem. I can't update to the latest version (2.11.0) because AppSync pod spec depends on AWSCore 2.10.2 insted of AWSAppSync 2.11.0

**Comment by JesusMartinAlonso** [comment]
I have detected a bug while using the new initializer released in 2.11.0. I have opened an issue #1939

**Pull request merged**